### ISSUE TITLE
Burn fusion optimisation

### ISSUE THREAD

**Issue by cBournhonesque** [tracel-ai/burn#2129]
Looking at the tests: https://github.com/cBournhonesque/burn/blob/d770b1f470d734afd79253d082ea9d084a8d5431/crates/burn-fusion/src/stream/execution/tests.rs#L89
it seems like if a segment doesn't find any optimization, the entire segment is executed with no fusion.

But wouldn't it be possible to just execute the first operation and then restart the fusion exploration on the second one?

On the test it looks like another optimization is potentially possible if [op_2] is added in the queue.

**Comment by nathanielsimard** [comment]
Yes, you are right! In fact, the real fix would be to have a start trigger for optimizations instead of always trying them out. It would be more natural and would avoid some bugs that we would have right now if we had more than just one optimization.

For now we're not missing any optimization because all element-wise operations are "start triggers" and all other ops are "stop trigger".

**Comment by antimora** [comment]
The fusion engine was rewritten since. Closing as outdated.
